### PR TITLE
Atomic bomb commit, should be feature complete.

### PR DESCRIPTION
**Pull request by MaxMahem**
Feature complete version of Manual Promotions, with the following exciting features!

* Two new options! One to turn Manual Promotions on, and one to turn Automatic Promotions off. The features can be used together or separately as desired.
* A new screen for performing manual promotions! Accessed by clicking on the rank icon.
* All available ranks for a soldier type are displayed.
* Availability slots are calculated based on mod rules.
* Ranks that cannot be promoted to are displayed in red.
* Ranks that have an unlimited number of promotion slots (Rookie and Squaddie) are notated properly.
* Ratios for different command slots are enforced in both promotion and demotion (cannot promote *or demote* to a full slot).
* Promotion rule for the commander is enforced. There can only be one!
* Quick search is functional (though rather pointless at the moment).
* Middle-clicking on the rank brings up the UFOpedia article, if any. (Demo articles not included in pull, but available if desired).

A demo: [https://www.youtube.com/watch?v=B13x_ptE97I&feature=youtu.be](url)

---

I am still pretty new at C++, but I tried my best not to emulate the style present and not step on any toes, though I did end up ripping into some existing functions a fair amount however. None of this code is in any performance-sensitive paths, but I tried to follow best practices as best I could. The code is fairly well tested in any case.

Less than ideal things I did, though:
* enums are used as index keys at times. Ideally, these would be replaced with a new `Rank` class and collection at some point in the future (maybe as part of the next patch), but I was aiming for minimal invasion. 
* Most of the logic is hard-coded in. Moving this logic to its own YAML rules is an obvious next step, but some design decisions will need to be made here (in particular, should rank rules be global, or vary by soldier type). 
* The code also has to deal with the fact that by default, the Rank Strings (i.e., `STR_ROOKIE`, `STR_SQUADDIE`, etc...) may be empty and provide default behavior in the case, which leads to some clunky code. Ideally, the default set of Rank Strings would be encoded in a YAML rule instead. Doing so might be a breaking change for mods that don't derive from standard rulesets (where the YAML values could be provided) or for mods with new soldier types but don't define these values. I am sure there is a good solution for this, but it's beyond the scope of what I wanted to do here.
* A lot of the promotion code is in `SaveGame` for some reason, despite having no dependencies. I moved some of it out, but the bulk of it remains. It should probably go into it's own file. Again, something for a later project, I think.

Anyways, this was fun to do. Thanks for consideration, and any feedback would be great!

### CODE REVIEW

**Comment by MaxMahem** [comment]
After some thought, I decided the fact that you cannot promote back up from rookie again after being demoted was too surprising, so I just disabled the ability to demote down to rookie entirely. I could make this configurable, but it's pretty minor, and honestly, I forsee soon making the whole system more configurable, which will make such an option obsolete.

File: src/Basescape/SoldierRankState.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	Base *_base;
+	size_t _soldierId;
+
+	TextButton *_btnCancel;
+	Window *_window;
+	Text *_txtTitle, *_txtRank, *_txtOpening;
+	TextEdit *_btnQuickSearch;
+	TextList *_lstRank;
+	std::vector<RankItem> _ranks;
+	std::vector<int> _indices;
```
Comment (MeridianOXC) [inline src/Basescape/SoldierRankState.h:60]: not needed, there is no sorting or filtering

File: src/Basescape/SoldierRankState.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+ */
+SoldierRankState::SoldierRankState(Base *base, size_t soldierId) : _base(base), _soldierId(soldierId)
+{
+	_screen = false;
+
+	// Create objects
+	_window = new Window(this, 192, 160, 64, 20, POPUP_BOTH);
+	_btnQuickSearch = new TextEdit(this, 48, 9, 80, 43);
+	_btnCancel = new TextButton(140, 16, 90, 156);
+	_txtTitle = new Text(182, 16, 69, 28);
```
Comment (MeridianOXC) [inline src/Basescape/SoldierRankState.cpp:52]: height 17 is needed, 16 is not enough

File: bin/standard/xcom2/interfaces.rul [hunk 3]
```diff
@@ -1557,6 +1557,21 @@ interfaces: #done
         border: 1  # white
       - id: button
         color: 1   # white
+  - type: soldierRank
+    backgroundImage: BACK14.SCR
+    palette: PAL_BASESCAPE
```
Comment (MeridianOXC) [inline bin/standard/xcom2/interfaces.rul:1562]: TFTD uses a different palette for basescape UIs, not PAL_BASESCAPE

File: bin/standard/xcom2/interfaces.rul [hunk 4]
```diff
@@ -1557,6 +1557,21 @@ interfaces: #done
...
+    elements:
+      - id: palette
+        color: 4     # green
+      - id: window
+        color: 218   # blue
+      - id: text
+        color: 213   # gold
+      - id: list
+        color: 208   # white
```
Comment (MeridianOXC) [inline bin/standard/xcom2/interfaces.rul:1571]: values are wrong and color2 is missing

File: bin/standard/xcom1/interfaces.rul [hunk 5]
```diff
@@ -1532,6 +1532,22 @@ interfaces:
...
+    elements:
+      - id: palette
+        color: 4     # green
+      - id: window
+        color: 218   # blue
+      - id: text
+        color: 213   # gold
+      - id: list
+        color: 208   # white
+        color2: 241  # pink
```
Comment (MeridianOXC) [inline bin/standard/xcom1/interfaces.rul:1547]: I will use 246 purple

File: src/Basescape/SoldierRankState.h (hunk 1 above)
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.h:60]: There is filtering, but not sorting.

File: src/Battlescape/DebriefingState.cpp [hunk 6]
```diff
@@ -81,7 +81,7 @@ namespace OpenXcom
  * Initializes all the elements in the Debriefing screen.
  * @param game Pointer to the core game.
  */
-DebriefingState::DebriefingState() : _eventToSpawn(nullptr), _region(0), _country(0), _positiveScore(true), _destroyBase(false), _showSellButton(true), _initDone(false), _pageNumber(0)
+DebriefingState::DebriefingState() : _eventToSpawn(nullptr), _region(0), _country(0), _positiveScore(true), _destroyBase(false), _showSellButton(true), _initDone(false), _pageNumber(0), _promotions(false)
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Battlescape/DebriefingState.cpp:84]: fixed, though pageNumber is mixed down with the methods which is weird. Left that alone.

File: src/Mod/Mod.h [hunk 7]
```diff
@@ -27,6 +27,7 @@
 #include "../Engine/FileMap.h"
 #include "../Engine/Collections.h"
 #include "../Savegame/GameTime.h"
+#include "../Savegame/Soldier.h"
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Mod/Mod.h:30]: yeah I think I mentioned this in my commit. I didn't like where all the code was but didn't want to radically change it all about right then, which meant some less than ideal things.

File: src/Savegame/Soldier.cpp [hunk 8]
```diff
@@ -746,6 +746,26 @@ void Soldier::promoteRank()
...
+ * Promotes a soldier to a specific rank (possibly a demotion)
+ */
+void Soldier::promoteRank(const SoldierRank newRank)
+{
+	if (!_rules->getAllowPromotion())
+		return;
+
+	const std::vector<std::string> &rankStrings = _rules->getRankStrings();
+	// abort if the desired rank is not indexed in the rank strings.
+	if (_rank >= rankStrings.size() - 1)
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/Soldier.cpp:759]: I'm not sure what you mean by newRank here, but I don't think this is off by one. The enum's are 0 index, so the size of the rankStrings should be one less.

File: src/Savegame/SavedGame.cpp [hunk 9]
```diff
@@ -2321,100 +2322,69 @@ Soldier *SavedGame::getSoldier(int id) const
...
-				soldierData.totalSergeants--;
-			}
-			else
-			{
-				break;
-			}
+			// only promote one colonel to commander
+			highestRanked->promoteRank();
+			soldiersPromoted++;
+			rankCounts[RANK_COMMANDER];
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/SavedGame.cpp:2372]: typo, fixed.

File: src/Savegame/RankCount.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	// can't promote to same rank.
+	if (currentRank == newRank)
+	{
+		return false;
+	}
+
+	// If the rankString for the soldier is not empty, check if the new rank is defined in the rank strings.
+	// If not, it is not allowed. If no rankStrings are defined, we assume default behavior.
+	const size_t rankStringsSize = soldierRules->getRankStrings().size();
+	if (rankStringsSize != 0 && newRank <= rankStringsSize - 1)
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/RankCount.cpp:118]: Ranges make my head hurt. Not sure how I missed this in testing but testing again it seems I did.

File: src/Basescape/SoldierRankState.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+#include "../Engine/State.h"
+#include "../SaveGame/Soldier.h"
+#include <vector>
+
+namespace OpenXcom
+{
+
+struct RankItem
+{
+	RankItem(const SoldierRank _rank, const std::string &_name, const int &_quantity, const bool _promotionAllowed)
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.h:29]: Sorry getting my feet under me here. Those members probably shouldn't be `const` as well.

File: src/Basescape/SoldierRankState.h [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+class SoldierRankState : public State
+{
+private:
+	Base *_base;
+	size_t _soldierId;
+
+	TextButton *_btnCancel;
+	Window *_window;
+	Text *_txtTitle, *_txtRank, *_txtOpening;
+	TextEdit *_btnQuickSearch;
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.h:57]: I left it in as a forward looking feature. Maybe someday ranks will be customizable and there might be more of them? Probably still not necessary but I thought it was fun.

File: src/Basescape/SoldierRankState.cpp [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+	_txtRank = new Text(90, 9, 80, 52);
+	_txtOpening = new Text(70, 9, 190, 52);
+	_lstRank = new TextList(160, 80, 73, 68);
+
+	// Set palette
+	setStandardPalette("PAL_BATTLESCAPE");
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.cpp:58]: misunderstood how that worked.

File: src/Basescape/SoldierRankState.cpp [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+			{
+				continue;
+			}
+		}
+
+		// if the promotion is not allowed, prepend a color flip to indicate this.
+		std::string rankText;
+		if (!rankItem.promotionAllowed)
+		{
+			rankText += Unicode::TOK_COLOR_FLIP;
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.cpp:162]: heh, didn't know that was an option!

File: bin/common/Language/en-US.yml [hunk 15]
```diff
@@ -283,6 +283,10 @@ en-US:
   STR_ALIENBLEEDING_DESC: "Allows fatal wounds to be inflicted on most aliens."
   STR_FIELDPROMOTIONS: "Field promotions"
   STR_FIELDPROMOTIONS_DESC: "Only soldiers that were present at the mission site are eligible for promotion."
+  STR_MANUALPROMOTIONS: "Manual promotions"
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply bin/common/Language/en-US.yml:286]: fixed

File: bin/common/Language/en-US.yml [hunk 16]
```diff
@@ -283,6 +283,10 @@ en-US:
...
+  STR_MANUALPROMOTIONS_DESC: "Soldiers can be manualy promoted or demoted by clilcking on the rank icon."
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply bin/common/Language/en-US.yml:287]: fixed

File: src/Engine/Options.cpp [hunk 17]
```diff
@@ -188,6 +188,8 @@ void create()
 	_info.push_back(OptionInfo("canTransferCraftsWhileAirborne", &canTransferCraftsWhileAirborne, false, "STR_CANTRANSFERCRAFTSWHILEAIRBORNE", "STR_GEOSCAPE")); // When the craft can reach the destination base with its fuel
 	_info.push_back(OptionInfo("retainCorpses", &retainCorpses, false, "STR_RETAINCORPSES", "STR_GEOSCAPE"));
 	_info.push_back(OptionInfo("fieldPromotions", &fieldPromotions, false, "STR_FIELDPROMOTIONS", "STR_GEOSCAPE"));
+	_info.push_back(OptionInfo("manualPromotions", &manualPromotions, false, "STR_MANUALPROMOTIONS", "STR_GEOSCAPE"));
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Engine/Options.cpp:191]: fixed

File: src/Engine/Options.inc.h [hunk 18]
```diff
@@ -23,7 +23,7 @@ OPT SDLKey keyOk, keyCancel, keyScreenshot, keyFps, keyQuickLoad, keyQuickSave;
 OPT int geoClockSpeed, dogfightSpeed, geoScrollSpeed, geoDragScrollButton, geoscapeScale;
 OPT bool includePrimeStateInSavedLayout, anytimePsiTraining, weaponSelfDestruction, retainCorpses, craftLaunchAlways, globeSeasons, globeDetail, globeRadarLines, globeFlightPaths, globeAllRadarsOnBaseBuild,
 	storageLimitsEnforced, canSellLiveAliens, canTransferCraftsWhileAirborne, customInitialBase, aggressiveRetaliation, geoDragScrollInvert,
-	allowBuildingQueue, showFundsOnGeoscape, psiStrengthEval, allowPsiStrengthImprovement, fieldPromotions, meetingPoint;
+	allowBuildingQueue, showFundsOnGeoscape, psiStrengthEval, allowPsiStrengthImprovement, fieldPromotions, manualPromotions, automaticPromotions, meetingPoint;
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Engine/Options.inc.h:26]: fixed

File: src/Mod/Mod.h [hunk 19]
```diff
@@ -899,14 +900,8 @@ class Mod
...
 	int getPilotsEmergencyEvacuationSurvivalChance() const { return _pilotsEmergencyEvacuationSurvivalChance; }
-	/// Gets how many soldiers are needed for one sergeant promotion
-	int getSoldiersPerSergeant() const { return _soldiersPerSergeant; }
-	/// Gets how many soldiers are needed for one captain promotion
-	int getSoldiersPerCaptain() const { return _soldiersPerCaptain; }
-	/// Gets how many soldiers are needed for one colonel promotion
-	int getSoldiersPerColonel() const { return _soldiersPerColonel; }
-	/// Gets how many soldiers are needed for one commander promotion
-	int getSoldiersPerCommander() const { return _soldiersPerCommander; }
+	/// Gets how many soldiers are needed for a given rank
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Mod/Mod.h:903]: fixed

File: src/OpenXcom.2010.vcxproj.user [hunk 20]
```diff
@@ -20,4 +20,7 @@
     <DebuggerFlavor>WindowsLocalDebugger</DebuggerFlavor>
     <LocalDebuggerCommandArguments>-data "$(ProjectDir)..\bin"</LocalDebuggerCommandArguments>
   </PropertyGroup>
+  <PropertyGroup>
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/OpenXcom.2010.vcxproj.user:23]: vs is giving me some grief sorry about that.

File: src/Savegame/Soldier.cpp [hunk 21]
```diff
@@ -746,6 +746,26 @@ void Soldier::promoteRank()
...
+	{
+		return;
+	}
+
+	_rank = newRank;
+
+	// note that we never need to show a notification for this style of promtion.
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/Soldier.cpp:766]: fixed

File: src/Savegame/Soldier.cpp [hunk 22]
```diff
@@ -746,6 +746,26 @@ void Soldier::promoteRank()
 	}
 }
 
+/**
+ * Promotes a soldier to a specific rank (possibly a demotion)
+ */
+void Soldier::promoteRank(const SoldierRank newRank)
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/Soldier.cpp:752]: renamed to changeRange

File: src/Savegame/RankCount.h [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+ * @brief Container for counts of soldier ranks.
+ */
+class RankCount : public RankCountBase
+{
+  public:
+	RankCount(const std::vector<Soldier *> soldiers);
+};
+
+/**
+ * @brief Container for count of promtion openings.
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/RankCount.h:53]: fixed

File: src/Savegame/RankCount.h [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+ */
+class PromotionOpenings : public RankCountBase
+{
+  public:
+	PromotionOpenings(const std::vector<Soldier *> soldiers, const Mod *mod);
+
+	// check if a soldier can be promoted to a give rank.
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/RankCount.h:60]: fixed

File: src/Savegame/RankCount.cpp [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+		_rankCounts[soldier->getRank()]++;
+	}
+	_totalSoldiers = soldiers.size();
+}
+
+/**
+ * Creates a new PromotionOpenings object, containing the count of promotion openings based on the mods promotion rules.
+ * Ranks that are overcomplement for some reason will show as 0 openings, ranks that have unlimited openings will have -1
+ * openings.
+ * @param soldiers the set of soldiers to consider for calcuation, typically should be all active soldiers.
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/RankCount.cpp:62]: fixed

File: src/Savegame/RankCount.cpp [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * @brief Tests if a soldier can be manually promoted to a new rank.
+ * @param soldier The soldier to test for promotion.
+ * @param newRank The new rank to promote the soldier to.
+ * @return True if the soldier can be promoted to this rank, false otherwise.
+ */
+const bool PromotionOpenings::isManualPromotionPossible(const Soldier *soldier, const SoldierRank newRank) const
+{
+	// check if the soldiers rules allow promotion.
+	auto const soldierRules = soldier->getRules();
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/RankCount.cpp:95]: fixed

File: src/Savegame/RankCount.cpp [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+	if (!soldierRules->getAllowPromotion())
+	{
+		return false;
+	}
+
+	const SoldierRank currentRank = soldier->getRank();
+
+	// rookies cannot be promoted or demoted to via this method.
+	if (currentRank == RANK_ROOKIE || newRank == RANK_ROOKIE)
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Savegame/RankCount.cpp:104]: What I would like is for the criteria to be more configurable, but that's a later discussion I guess.

File: src/Basescape/SoldierRankState.h [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+	bool promotionAllowed;
+};
+
+class Base;
+class TextButton;
+class Window;
+class Text;
+class TextEdit;
+class TextList;
+class Armor;
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.h:45]: removed

File: src/Basescape/SoldierRankState.h [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+	/// togges the quick search visible and not
+	void btnQuickSearchToggle(Action *action);
+	/// applies the quick search
+	void btnQuickSearchApply(Action *);
+	/// Handler for clicking the Cancel button.
+	void btnCancelClick(Action *action);
+	/// Handler for clicking the Rank list (promotes if possible).
+	void lstRankClick(Action *action);
+	/// Handler for middle clicking the Rank list (ufopedia article if possible).
+	void lstArmorClickMiddle(Action *action);
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.h:76]: fixed

File: src/Basescape/SoldierRankState.h [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+struct RankItem
+{
+	RankItem(const SoldierRank _rank, const std::string &_name, const int &_quantity, const bool _promotionAllowed)
+		: rank(_rank), name(_name), openings(_quantity), promotionAllowed(_promotionAllowed)
+	{
+	}
+	const SoldierRank rank;
+	const int openings;
+	const std::string name;
+	bool promotionAllowed;
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.h:36]: welp. Fixed.

File: src/Basescape/SoldierRankState.cpp [hunk 31]
```diff
@@ -0,0 +1,244 @@
...
+	std::vector<std::string> rankStrings = soldierRules->getRankStrings();
+
+	// if the rank strings is empty and the soldier is allowed promotion, build a default set of rank strings.
+	// ideally this would be encoded in the data someplace instead of relying on this behavior for empty rank strings.
+	if (rankStrings.empty() && soldierRules->getAllowPromotion())
+	{
+		rankStrings = {"STR_ROOKIE", "STR_SQUADDIE", "STR_SERGEANT", "STR_CAPTAIN", "STR_COLONEL", "STR_COMMANDER"};
+	}
+
+	SoldierRank currentRank = soldier->getRank();
```
[the remark this answers is not in the document]
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.cpp:116]: removed.

File: src/Basescape/SoldierRankState.cpp (hunk 2 above)
Comment (MaxMahem) [reply src/Basescape/SoldierRankState.cpp:52]: increased.

**Comment by MeridianOXC** [comment]
Manually merged: ae0e4d16393f133eb3217c286db4db9e5109c25b

**Pull request closed without merging**